src/mxnet/Symbol.d

```
@@ -860,6 +915,29 @@ public class SoftmaxOutput : Symbol
     }
 }
 
+///
+unittest
+{
+    // set up input and label variables
+    scope input = new Variable("x");
+    scope (exit) input.freeHandle();
+    scope label = new Variable("y");
+    scope (exit) label.freeHandle();
+
+    // create SoftmaxOutput symbol using default configuration
+    scope softmax_default = new SoftmaxOutput(input, label);
+    scope (exit) softmax_default.freeHandle();
+
+    // create SoftmaxOutput symbol with custom configuration
+    SoftmaxOutput.Config config = {
+        grad_scale: 1.5,
```

I would suggest here to explicitly write `1.5f`.

---

Is there a particular reason?  It doesn't make any difference, and in general I'd resist qualifying numbers in this way unless it's strictly _necessary_.  Otherwise it just risks creating annoyances in future (if e.g. the floating-point type used changes).

---

When reading this line, I would now assume that double precision is used. But in fact only single precision is because this is (silently) converted to a `float` value.

---

Well, it's a floating point value.  There's no harm in specifying it in default precision, knowing very well that it may be stored in something smaller.

Switch the situation around: suppose that the underlying implementation changes to `double`, now we have to (remember to) rewrite the call site as well?

---

Basically, the suggestion here is that we avoid unnecessary use of qualifiers for numerical constants, because it creates an extra maintenance burden of things to rewrite _in the case that the type changes_, for not a lot of benefit in terms of readability.

The only situation where we should be worrying about that is where it actually _matters_ to specifically have a `float` (e.g. because it's necessary to match exact test results).

---

Note though that sometimes the value cannot be converted properly. This is rare but it can happen. But here nothing should go wrong. I tend to lean more toward explicitly. In D a `float` is silently converted to a `double` which looks odd to me. But I understand your reasoning.